### PR TITLE
Issue #3415: Views: Incorrect filter group OR behavior, LEFT JOIN cha…

### LINKED ISSUE #3415: Views: Incorrect filter group OR behavior, LEFT JOIN changed to INNER JOIN
This is the respective issue for https://www.drupal.org/project/drupal/issues/1766338

That issue with Views seems to have regressed between 3.3 (released Feb 2012) and 3.4 (Aug 2012). Given that Backdrop 1.0.0 was released in Jan 2015, we must have inherited it.

As a summary, this issue is causing views with a SQL query containing a logical structure as simple as the following to not work correctly:
```
  AND A, B
OR
  AND X, Y
```
Given such a basic structure, it's likely that a huge number of views are not working correctly.

There is [a comment](https://www.drupal.org/project/drupal/issues/1766338#comment-6424788) in the issue in d.org which is pointing to a specific commit (`994acb299405ff0651706c020cf5d4e6d689e8b3`), which seems to only include a class name change. So it seems that the commit was part of [issue 1494884](https://www.drupal.org/project/views/issues/1494884), which is currently in "needs work" status.

Steps to reproduce:

* to a content type with existing content: add a boolean field
* to a view for published content of that type: add the new field to the filter as OR - Is one of - 1
* Expected: No change on the result list
* Actual: No results

---

~PR by @klonos https://github.com/backdrop/backdrop/pull/2414~
~PR by @jenlampton https://github.com/backdrop/backdrop/pull/3100~
PR by @laryn https://github.com/backdrop/backdrop/pull/3702

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/3415

Replaces https://github.com/backdrop/backdrop/pull/2414

### CODE REVIEW

**Comment by stpaultim** [comment]
@jenlampton I think the sandbox still needs refreshing?

**Comment by laryn** [comment]
@jenlampton I tweaked your PR in a way that is now working for me -- here's the relevant section. Does that make sense to you? (It wasn't working and I was also getting a lot of `Notice: Undefined index: plugin_id in views_many_to_one_helper->ensure_my_table() (line 796...` in my logs).

```
          if (isset($join)) {
            $join->type = 'LEFT';
            // INNER joins are slightly faster, use them when we know its safe.
            $group = isset($this->handler->options['group']) ? $this->handler->options['group'] : FALSE;
            $group_type = $this->handler->query->where[$group]['type'];
            $plugin_id = isset($this->handler->options['plugin_id']) ? $this->handler->options['plugin_id'] : FALSE;
            if ($group === FALSE || $group_type == 'AND' && $plugin_id && $plugin_id != 'taxonomy_index_tid' ) {
              $join->type = 'INNER';
            }
          }

```

**Comment by jenlampton** [comment]
This looks great, can you file a new PR? I'd be happy to close this one in favor of yours.

**Comment by laryn** [comment]
@jenlampton Okay, I just filed a new PR. Some random unrelated commits snuck in when I tried to refresh my fork... but they don't change the code and squash and merge should take care of them, I think?

 - https://github.com/backdrop/backdrop/pull/3702

File: core/modules/views/includes/handlers.inc [hunk 1]
```diff
@@ -785,11 +785,16 @@ class views_many_to_one_helper {
...
-          // when we know we can.
           $join = $this->get_join();
           if (isset($join)) {
-            $join->type = 'INNER';
+            $join->type = 'LEFT';
+            // INNER joins are slightly faster, use them when we know its safe.
+            $group = isset($this->handler->options['group']) ? $this->handler->options['group'] : FALSE;
+            $group_type = $this->handler->query->where[$group]['type'];
+            $plugin_id = $this->handler->options['plugin_id'];
+            if ($group === FALSE || $group_type == 'AND' && $plugin_id != 'taxonomy_index_tid' ) {
```
Comment (klonos) [inline core/modules/views/includes/handlers.inc:795]:
```suggestion
            if ($group === FALSE || $group_type == 'AND' && $plugin_id != 'taxonomy_index_tid') {
```

**Comment by laryn** [comment]
Extended by https://github.com/backdrop/backdrop/pull/3702

Closing per @jenlampton comment above: 

> This looks great, can you file a new PR? I'd be happy to close this one in favor of yours.

**Pull request closed without merging**